compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt

```diff
@@ -5,32 +5,132 @@
-            if (member is FirProperty) {
-                checkProperty(declaration, member, context, reporter)
+        val memberProperties = declaration.declarations.filterIsInstance<FirProperty>()
+
+        // A property is known to be initialized only if it is initialized
+        //   1) with its own initializing expression;
+        //   2) at every class constructor;
+        //   3) at any of class's anonymous initializers; or
+        //   4) at other property's initializing expression
+
+        // 2) Property can be initialized at constructors. Since it's unknown what constructor will be used, the property can be determined
+        // as initialized only if it is initialized at every constructor. The caveat is a constructor that calls another constructor, e.g.,
+        //   constructor() { x = ... }
+        //   constructor(...): this() { ... }  // x will be initialized via this() delegation
+
+        // 3) Property can be initialized at any of class's anonymous initializers (all of initializers will be executed), e.g.,
+        //   init { x = ... }
+        //   ...
+        //   init { y = ... }
+
+        // 4) Property can be initialized at other property's initializing expression too, e.g.,
+        //   val initX = inlineMe { x = ... } // where inlineMe returns the value of the last expression of the lambda
+
+        // Here, we accumulate initialized info of a property per constructor, anonymous initializer, or property initialization
+        val initializedInConstructor = mutableMapOf<FirProperty, EventOccurrencesRange>().withDefault { EventOccurrencesRange.ZERO }
+        val initializedInInit = mutableMapOf<FirProperty, EventOccurrencesRange>().withDefault { EventOccurrencesRange.ZERO }
+        val initializedInOtherProperty = mutableMapOf<FirProperty, EventOccurrencesRange>().withDefault { EventOccurrencesRange.ZERO }
+
+        // To handle the delegated constructor call, we need a cache from constructor to (analyzed) property init info.
+        val constructorToData = mutableMapOf<FirConstructor, PropertyInitializationInfo>().withDefault { PropertyInitializationInfo.EMPTY }
+
+        fun collectInfoFromGraph(
+            graph: ControlFlowGraph,
+            map: MutableMap<FirProperty, EventOccurrencesRange>,
+            acc: (EventOccurrencesRange, EventOccurrencesRange) -> EventOccurrencesRange,
+            delegatedConstructor: FirConstructor? = null,
+        ) {
+            val delegatedInfo =
+                if (delegatedConstructor != null && constructorToData.containsKey(delegatedConstructor)) {
+                    constructorToData.getValue(delegatedConstructor)
+                } else PropertyInitializationInfo.EMPTY
+
+            val data = PropertyInitializationInfoCollector(memberProperties.map { it.symbol }.toSet()).getData(graph)
+            val infoAtExitNode = data[graph.exitNode]?.get(NormalPath) ?: PropertyInitializationInfo.EMPTY
+
+            val info = when {
```

Comment:
They will be same, yet unnecessarily expensive. I rather move the logic to check either side is empty to `...Range` and made here simpler. We can't do the same for `merge`, though, since seeing `ZERO` on one side still matters.